scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+# 0001 - ScannerV4 APIs definition
+
+- **Status:** Accepted
+- **Created:** [2023-05-09 Tue]
+
+## Context
+
+We are transitioning Stackrox Scanner to use [ClairCore](https://github.com/quay/claircore) as its underlying scanning engine. The goal is to align vulnerability scanning results between Quay/Clair and Stackrox. The transition will be gradual, starting with container image scanning. During this period, the existing Scanner (ScannerV2) and the new ClairCore-based Scanner (ScannerV4) will be simultaneously deployed.
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using a "hash_id" managed by the clients. It uniquely identifies the resource's manifest.
+
+ScannerV4 will replace ScannerV2 as the container image scanner in Central. A new image integration for ScannerV4 will be created and replace ScannerV2's integration (type "clarify"), activated by the feature flag: `ROX_SCANNER_V4_ENABLED`. The "clarify" type will continue to be used for orchestrator and node scanning.
+
+ScannerV4 gRPC endpoints will use GRPC status codes to communicate the return status in case of non-successful responses.
+
+ScannerV4 will also provide endpoints for health and observability checks, which are not detailed in this ADR.
+
+### ScannerV4 API Endpoints
+
+The endpoints are versioned at `v4` to align with Clair:
```

Comment:
You'll need a `scanner.v4.Indexer/DeleteIndex` to prevent the backing database from growing unbounded.

Reply:
Who would own the lifecycle of IndexerReports? Right now, I think Scanner would take care of that without exposing that complexity to clients (Central and Sensor). But I would like to know more. Can you share some insights on how it is done in Clair?

Reply:
Clair's meant to be used with a registry, so the registry owns the lifecycle of the manifest. Having to deal with Clair possibly deleting the manifest out from underneath the registry makes the client much more complicated just to avoid keeping track of something it has to keep track of anyway.

Reply:
> Having a delete makes sense, especially because we will be persisting the data, unlike before. 

Consider removing unused manifests periodically instead of relying on a Delete function. Clients won't need to manage the Index lifecycle. I think it's best to exclude the delete option unless we are going to manage it independently of Scanner.

Reply:
> Maybe we could consider deleting index reports upon deployment deletion (so, put it on Central/Sensor to call a Delete endpoint)

Deployments are more volatile than images stored in registries. It might be we will perform deletion and re-creation more often than ideal. Also, hard to establish if a single deployment is the last reference to an image in the cluster.

Reply:
I will leave deletion out of the decision. We already discussed how Scanner would own the database storage. Moving forward, we can simply introduce it if needed and supersede this ADR.